hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestCompactingMemStore.java

```diff
@@ -0,0 +1,1391 @@
+    long nanos = System.nanoTime();
+    KeyValueScanner s = ms.getScanners(0).get(0);
+    s.seek(KeyValueUtil.createFirstOnRow(new byte[] { }));
+
+    System.out.println(iteration + " create/seek took: " + (System.nanoTime() - nanos) / 1000);
+    int cnt = 0;
+    while (s.next() != null) ++cnt;
+
+    System.out.println(iteration + " took usec: " + (System.nanoTime() - nanos) / 1000 + " for: "
+        + cnt);
+
+  }
+
+  @Override
+  public void tearDown() throws Exception {
+    chunkPool.clearChunks();
+  }
+
+  @Override
+  public void setUp() throws Exception {
+    super.setUp();
+    this.mvcc = new MultiVersionConcurrencyControl();
+    Configuration conf = new Configuration();
+    conf.setBoolean(SegmentFactory.USEMSLAB_KEY, true);
+    conf.setFloat(MemStoreChunkPool.CHUNK_POOL_MAXSIZE_KEY, 0.2f);
+    conf.setInt(HRegion.MEMSTORE_PERIODIC_FLUSH_INTERVAL, 1000);
+    HBaseTestingUtility hbaseUtility = HBaseTestingUtility.createLocalHTU(conf);
+    HColumnDescriptor hcd = new HColumnDescriptor(FAMILY);
+    this.region = hbaseUtility.createTestRegion("foobar", hcd);
+    this.regionServicesForStores = region.getRegionServicesForStores();
+    this.store = new HStore(region, hcd, conf);
+    this.cms = new CompactingMemStore(HBaseConfiguration.create(), CellComparator.COMPARATOR,
+        store, regionServicesForStores);
+    chunkPool = MemStoreChunkPool.getPool(conf);
+    assertTrue(chunkPool != null);
+  }
+
+  public void testPutSameKey() {
+    byte[] bytes = Bytes.toBytes(getName());
+    KeyValue kv = new KeyValue(bytes, bytes, bytes, bytes);
+    this.cms.add(kv);
+    byte[] other = Bytes.toBytes("somethingelse");
+    KeyValue samekey = new KeyValue(bytes, bytes, bytes, other);
+    this.cms.add(samekey);
+    Cell found = this.cms.getActive().first();
+    assertEquals(1, this.cms.getActive().getCellsCount());
+    assertTrue(Bytes.toString(found.getValueArray()), CellUtil.matchingValue(samekey, found));
+  }
+
+  /**
+   * Test memstore snapshot happening while scanning.
+   *
+   * @throws IOException
+   */
+  public void testScanAcrossSnapshot() throws IOException {
+    int rowCount = addRows(this.cms);
+    List<KeyValueScanner> memstorescanners = this.cms.getScanners(0);
+    Scan scan = new Scan();
+    List<Cell> result = new ArrayList<Cell>();
+    Configuration conf = HBaseConfiguration.create();
+    ScanInfo scanInfo =
+        new ScanInfo(conf, null, 0, 1, HConstants.LATEST_TIMESTAMP, KeepDeletedCells.FALSE, 0,
+            this.cms.getComparator());
+    ScanType scanType = ScanType.USER_SCAN;
+    StoreScanner s = new StoreScanner(scan, scanInfo, scanType, null, memstorescanners);
+    int count = 0;
+    try {
+      while (s.next(result)) {
+        LOG.info(result);
+        count++;
+        // Row count is same as column count.
+        assertEquals(rowCount, result.size());
+        result.clear();
+      }
+    } finally {
+      s.close();
```

Comment:
Looks like all these tests are cut-and-paste. If this is indeed the case, let's discuss how to avoid duplication.